src/tfr_mission_control/src/tfr_mission_control/mission_control.cpp

```diff
@@ -0,0 +1,409 @@
+                performTeleop(tfr_utilities::TeleopCode::LEFT);
+            }
+            else if(k == Qt::Key_D)
+            {
+                motorKill->start(MOTOR_INTERVAL);
+                lastKey =k;
+                performTeleop(tfr_utilities::TeleopCode::RIGHT);
+            }
+            //consume the key
+            return true;
+        }
+        else {
+            //unrecognized event type, pass it on
+            return QObject::eventFilter(obj, event);
+        }
+    }
+
+    /* ========================================================================== */
+    /* Callbacks                                                                  */
+    /* ========================================================================== */
+    /*
+     * Callback for our status subscriber this happens in the "ros" thread so I
+     * need to communicate to the GUI in a safe signal/slot combination.
+     *
+     * This triggers our custom emitStatus signal, which triggers the built in
+     * text update in the text box, a seperate timer event is constantly
+     * scrolling the window on a set fast interval.
+     * */
+    void MissionControl::updateStatus(const tfr_msgs::SystemStatusConstPtr &status)
+    {
+        auto str = getStatusMessage(static_cast<StatusCode>(status->status_code), status->data);
+        //need to wrap in signal safe "Q" object
+        QString msg = QString::fromStdString(str);
+        emit emitStatus(msg);
+    }
+
+    /* ========================================================================== */
+    /* Slots                                                                      */
+    /* ========================================================================== */
+
+    //self explanitory, starts the time service in executive
+    void MissionControl::startTimeService()
+    {
+        tfr_msgs::EmptySrv start;
+        ros::service::call("start_mission", start);
+        //start updating the gui mission clock
+        countdown->start(500);
+    }
+
+    //starts mission in autonomous mode
+    void MissionControl::startMission()
+    {
+        startTimeService();
+        goAutonomousMode();
+    }
+    
+    //starts mission is teleop mode
+    void MissionControl::startManual()
+    {
+        startTimeService();
+        goTeleopMode();
+    }
+
+    //triggers state change into autonomous mode from teleop
+    void MissionControl::goAutonomousMode()
+    {
+        softwareStop();
+        setAutonomy(true);
+        tfr_msgs::EmptyGoal goal{};
+        while (!teleop.getState().isDone()) teleop.cancelAllGoals();
+        autonomy.sendGoal(goal);
+        setTeleop(false);
+        widget->setFocus();
+
+    }
+
+    //triggers state change into from teleop into autonomy
+    void MissionControl::goTeleopMode()
+    {
+        setAutonomy(false);
+        while (!autonomy.getState().isDone()) autonomy.cancelAllGoals();
+        setTeleop(true);
+        softwareStop();
```

Comment:
What's the reasoning for having softwareStop() so late in this method (vs. goAutonomousMode() where it's the first line)? I'm guessing it's for letting autonomy hand over control of the actuators first before we stop them?

Reply:
yes. Notice how we are letting the autonomous action server preempt itself? on line 364.

Reply:
Makes sense, yeah, if everything else also stops the motor.